File: EatSomewhere/Data/FoodEntry.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using EatSomewhere.Users;

namespace EatSomewhere.Data;

public class FoodEntry
{
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public string Id { get; set; }
    public DateTime Date { get; set; }
    public string? Comment { get; set; }
    public Food Food { get; set; }
    public long EstimatedCost => (long)decimal.Ceiling(Food.EstimatedCost / (decimal)Food.PersonCount * PersonCount);
    public long Cost { get; set; }

    public long CostPerPerson => (long)decimal.Ceiling(Cost / (decimal)Math.Max(1, PersonCount));
    
    public int PersonCount => Participants.Sum(x => x.Persons);
    public List<FoodParticipant> Participants { get; set; }
    public User? PayedBy { get; set; }
    public Assembly Assembly { get; set; }
    public User CreatedBy { get; set; }
    public List<Bill> Bills { get; set; }

    public List<Bill> CalculateBills()
    {
        List<Bill> bills = Participants.Select(x => new Bill
        {
            User = x.User ?? null,
            Amount = CostPerPerson * x.Persons,
            Persons = x.Persons,
            Recipient = PayedBy,
            Date = Date
        }).ToList();
        int i = 0;
        while (bills.Sum(x => x.Amount) > Cost)
        {
            bills[i % bills.Count].Amount -= 1;
            i++;
        }
        return bills;
    }
}

File: app/lib/data/foodentry.dart
import 'package:eat_somewhere/backend_data/Backend_user.dart';
import 'package:eat_somewhere/data/bill.dart';
import 'package:eat_somewhere/data/food.dart';

class FoodEntry {
  String? id;
  DateTime date = DateTime.now();
  String? comment;
  Food? food;
  int estimatedCost = 0;
  int cost = 0;
  int costPerPerson = 0;
  int personCount = 1;
  List<FoodParticipant> participants = [];
  List<Bill?> bills = [];
  BackendUser? payedBy;
  String? assemblyId;
  bool inProgress = false;

  int getPersonCount() {
    return participants.fold(0, (sum, participant) => sum + participant.getPersonAmount());
  }

  int getEstimatedCost() {
    if (food == null) return 0;
    return food!.getCostPerPerson() * getPersonCount();
  }

  FoodEntry.fromJson(Map<String, dynamic> json) {

    id = json["Id"];
    assemblyId = json["Assembly"]["Id"];
    date = DateTime.parse(json["Date"]);
    comment = json["Comment"];
    food = Food.fromJson(json["Food"]);
    inProgress = json["InProgress"] ?? false;
    cost = json["Cost"];
    costPerPerson = json["CostPerPerson"];
    personCount = json["PersonCount"];
    bills = (json["Bills"] as List<dynamic>)
        .map((e) => Bill.fromJson(e))
        .toList();
    if(bills == null) {
      bills = [];
    }
    participants = (json["Participants"] as List<dynamic>)
        .map((e) => FoodParticipant.fromJson(e))
        .toList();
    if(json["PayedBy"] != null) payedBy = BackendUser.fromJson(json["PayedBy"]);
  }

  FoodEntry();

  toJson() {
    inProgress = cost <= 0;
    return {
      "Assembly": {
        "Id": assemblyId,
      },
      "Id": id,
      "InProgress": inProgress,
      "Date": date.toIso8601String(),
      "Comment": comment,
      "Food": food?.toJson(),
      "Cost": cost,
      "Participants": participants.map((e) => e.toJson()).toList(),
      "PayedBy": payedBy?.toJson(),
    };
  }

  int getCostPerPerson() {
    if (cost == 0) {
      return 0;
    }
    if(getPersonCount() == 0) return 0;
    return (cost / getPersonCount()).ceil();
  }

  List<IngredientEntry> getShoppingList() {
    List<IngredientEntry> shoppingList = [];
    int persons = getPersonCount();
    if(food == null) return [];
    for (IngredientEntry ingredient in food!.ingredients) {
      print("Ingredient: ${ingredient.ingredient}");
      if(ingredient.ingredient == null) continue;
      IngredientEntry shoppingItem = IngredientEntry.fromIngredient(ingredient.ingredient!);
      shoppingItem.amount = ingredient.amount / (food?.personCount ?? 1) * persons;
      shoppingList.add(shoppingItem);
    }
    return shoppingList;
  }

  FoodEntry.fromFood(Food this.food);
}

class FoodParticipant {
  String? id;
  BackendUser? user;
  int additionalPersons = 0;

  int getPersonAmount() {
    return additionalPersons + 1;
  }

  FoodParticipant.fromJson(Map<String, dynamic> json) {
    id = json["Id"];
    user = BackendUser.fromJson(json["User"]);
    additionalPersons = json["AdditionalPersons"];
  }

  toJson() {
    return {
      "Id": id,
      "User": user?.toJson(),
      "AdditionalPersons": additionalPersons,
    };
  }

  FoodParticipant.fromUser(BackendUser backendUser) {
    user = backendUser;
    additionalPersons = 0;
  }
}